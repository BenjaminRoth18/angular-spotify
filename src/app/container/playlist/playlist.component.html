<div class="container">
    <div class="row">
        <div class="col-6 col-md-4 mb-5" *ngFor="let track of playlist">
            <div class="track">
                <img [src]="track.imageUrl" class="card-img-top" alt="" />
                <div class="track__info">
                    <h5 class="mb-1">{{ track.track }}</h5>
                    <p>by {{ track.artist }}</p>
                </div>
            </div>
        </div>
    </div>
</div>
