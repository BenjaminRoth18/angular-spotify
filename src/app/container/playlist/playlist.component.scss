img {
    border-radius: 0;
    position: relative;
}

.track {
    position: relative;

    &__info {
        background: rgb(0, 0, 0);
        background: linear-gradient(
            0deg,
            rgba(0, 0, 0, 0.8) 0%,
            rgba(0, 0, 0, 0) 100%
        );
        bottom: 0;
        color: white;
        padding: 2rem 1rem 0 1rem;
        position: absolute;
        width: 100%;

        h5 {
            font-size: 1rem;
        }

        p {
            font-size: 0.75rem;
        }
    }
}

@media (min-width: 1200px) {
    .track {
        position: relative;

        &__info {
            h5 {
                font-size: 1.25rem;
            }

            p {
                font-size: 1rem;
            }
        }
    }
}
