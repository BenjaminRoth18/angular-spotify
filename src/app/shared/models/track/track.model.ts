export interface Track {
    artist: string;
    track: string;
    imageUrl: string;
}

export interface PlaylistDto {
    tracks: {
        items: PlaylistItemDto[];
    };
}

export interface PlaylistItemDto {
    track: PlaylistTrackDto;
}

export interface PlaylistTrackDto {
    album: {
        images: Array<{ url: string }>;
    };
    artists: Array<{ name: string }>;
    name: string;
}
