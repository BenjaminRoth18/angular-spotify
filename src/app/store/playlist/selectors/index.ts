import { createSelector } from '@ngrx/store';

import * as fromPlaylistReducers from '../reducers/playlist.reducers';

export interface PlaylistAppState {
    playlist: fromPlaylistReducers.State;
}

export const selectPlaylist = (state: PlaylistAppState) => state.playlist;

export const selectPlaylistTracks = createSelector(
    selectPlaylist,
    (state: fromPlaylistReducers.State) => state.tracks
);
