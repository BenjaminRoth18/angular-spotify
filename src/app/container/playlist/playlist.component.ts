import { Component, OnDestroy, OnInit } from '@angular/core';
import { Subscription } from 'rxjs';
import { Track } from 'src/app/shared/models/track/track.model';
import { LayoutStateService } from 'src/app/store/layout/services/layout.state.service';
import { PlaylistStateService } from 'src/app/store/playlist/services/playlist.state.service';

@Component({
    selector: 'sp-playlist',
    templateUrl: './playlist.component.html',
    styleUrls: ['./playlist.component.scss']
})
export class PlaylistComponent implements OnInit, OnDestroy {
    playlist: Track[];
    subscriptions = new Subscription();
    loadingCount: number;

    constructor(
        private playlistStateService: PlaylistStateService,
        private layoutStateService: LayoutStateService
    ) {}

    ngOnInit(): void {
        this.subscriptions.add(
            this.playlistStateService
                .getPlaylist()
                .subscribe((playlist: Track[]) => {
                    this.playlist = playlist;
                })
        );

        this.subscriptions.add(
            this.layoutStateService
                .getLoadingCount()
                .subscribe((loadingCount: number) => {
                    this.loadingCount = loadingCount;
                })
        );
    }

    ngOnDestroy() {
        this.subscriptions.unsubscribe();
    }
}
