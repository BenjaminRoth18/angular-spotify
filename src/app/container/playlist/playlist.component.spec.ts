import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { of } from 'rxjs';
import { MockLoaderComponent } from 'src/app/shared/components/loader/loader.component.mock';
import { trackMock } from 'src/app/shared/models/track/track.mock';
import { LayoutStateService } from 'src/app/store/layout/services/layout.state.service';
import { MockLayoutStateService } from 'src/app/store/layout/services/layout.state.service.mock';
import { PlaylistStateService } from 'src/app/store/playlist/services/playlist.state.service';
import { MockPlaylistStateService } from 'src/app/store/playlist/services/playlist.state.service.mock';

import { PlaylistComponent } from './playlist.component';

describe('PlaylistComponent', () => {
    let component: PlaylistComponent;
    let fixture: ComponentFixture<PlaylistComponent>;
    let playlistStateService: PlaylistStateService;
    let layoutStateService: LayoutStateService;

    beforeEach(async(() => {
        TestBed.configureTestingModule({
            providers: [
                {
                    provide: PlaylistStateService,
                    useClass: MockPlaylistStateService
                },
                {
                    provide: LayoutStateService,
                    useClass: MockLayoutStateService
                }
            ],
            declarations: [PlaylistComponent, MockLoaderComponent]
        });

        fixture = TestBed.createComponent(PlaylistComponent);
        component = fixture.debugElement.componentInstance;
        playlistStateService = fixture.debugElement.injector.get(
            PlaylistStateService
        );
        layoutStateService = fixture.debugElement.injector.get(
            LayoutStateService
        );
    }));

    it('should create component', () => {
        expect(component).toBeTruthy();
    });

    it('should handle ngOnInit subscriptions', () => {
        playlistStateService.getPlaylist = jest
            .fn()
            .mockReturnValue(of([trackMock]));
        layoutStateService.getLoadingCount = jest.fn().mockReturnValue(of(1));
        component.ngOnInit();

        expect(component.playlist).toEqual([trackMock]);
        expect(component.loadingCount).toEqual(1);
    });

    describe('snaphots', () => {
        it('should create snapshot', () => {
            expect(fixture).toMatchSnapshot();
        });
    });
});
