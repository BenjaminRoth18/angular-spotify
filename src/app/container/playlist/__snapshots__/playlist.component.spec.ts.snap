// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistComponent snaphots should create snapshot 1`] = `
<sp-playlist
  layoutStateService={[Function MockLayoutStateService]}
  playlistStateService={[Function MockPlaylistStateService]}
  subscriptions={[Function Subscription]}
>
  <div
    class="container"
  >
    <div
      class="row"
    >
      
    </div>
  </div>
</sp-playlist>
`;
